website/docs/cli/commands/test/index.mdx

```diff
@@ -0,0 +1,368 @@
+        <CodeBlock language={"hcl"}>{ModuleTest}</CodeBlock>
+    </TabItem>
+    <TabItem value={"helper"} label={"test-helper/helper.tf"} default>
+        <CodeBlock language={"hcl"}>{ModuleHelper}</CodeBlock>
+    </TabItem>
+</Tabs>
+
+This project uses a [third-party provider](https://github.com/kreuzwerker/terraform-provider-docker) to launch the
+container. You can run it locally if you have a Docker Engine installed.
+
+:::
+
+### The `variables` and `run.variables` blocks
+
+The code under test (e.g. `main.tf`) will often have variable blocks that you need to fill from your test case. You
+can provide variables to your test run using any of the following methods:
+
+| Order  | Source                                                                                                                         |
+|:------:|:-------------------------------------------------------------------------------------------------------------------------------|
+|   1    | Environment variables with the `TF_VAR_` prefix.                                                                               |
+|   2    | tfvar files specified: `terraform.tfvars` and `*.auto.tfvars`.                                                                 |
+|   3    | Commandline variables defined using the flag `-var`, and the variables defined in the files specified by the flag `-var-file`. |
+|   4    | The variables from the `variables` block in a test file.                                                                       |
+|   5    | The variables from the `variables` block in `run` block.                                                                       |
+
+OpenTofu evaluates the variables in the order listed above, so you can use it to override the previously set variable.
+For example:
+
+<Tabs>
+    <TabItem value="test" label="main.tftest.hcl" default><CodeBlock language={"hcl"}>{VariablesTest}</CodeBlock></TabItem>
+    <TabItem value="main" label="main.tf"><CodeBlock language={"hcl"}>{VariablesMain}</CodeBlock></TabItem>
+</Tabs>
+
+### The `run.expect_failures` list
+
+In some cases you may want to test deliberate failures of your code, for example to ensure your validation is working.
+
+You can use the `expect_failures` inside a `run` block to specify which variables or resources should fail when the
+code is run with the given parameters.
+
+For example, the test case below checks if the `instances` variable correctly fails validation when supplied with a
+negative number:
+
+<Tabs>
+    <TabItem value={"test"} label={"main.tftest.hcl"} default>
+        <CodeBlock language={"hcl"}>{ExpectFailureVariablesTest}</CodeBlock>
+    </TabItem>
+    <TabItem value={"main"} label={"main.tf"}>
+        <CodeBlock language={"hcl"}>{ExpectFailureVariablesMain}</CodeBlock>
+    </TabItem>
+</Tabs>
+
+You can also use the `expect_failure` clause to check lifecycle events like pre- or postconditions as well as
+the results of checks.
```

Comment:
Maybe we can link here to http://opentofu.org/docs/language/meta-arguments/lifecycle/?